### ISSUE TITLE
t/test.pl:  Insecure dependency in require in taint mode

### ISSUE THREAD

**Issue by jkeenan** [atoomic/perl#11]
Background:  In my own code for $self, $job or CPAN, I never used taint mode.

Examples:

`t/run/switcht.t` warns but then runs:

```
Insecure dependency in require while running with -t switch at ./test.pl line 27.
run/switcht.t ..................... 
1..13
ok 1 - ${^TAINT} == -1
ok 2 - `` worked
ok 3 -     taint warn
ok 4 - `` worked
ok 5 -    no warnings "taint"
ok 6 - open >
ok 7 - open > taint warn
ok 8 -    file written
ok 9 - unlink() taint warn
ok 10 - unlink worked
ok 11 - -t doesn't enable regular warnings
ok 12 # skip skip tainted do test with \ separator
ok 13 - Could "do" ./ttdir/ttest.pl
ok
```
Four other files warn and then fail in a `BEGIN` block.
```
Insecure dependency in require while running with -T switch at ./test.pl line 27.
BEGIN failed--compilation aborted at ./test.pl line 29.
Compilation failed in require at op/taint.t line 12.
BEGIN failed--compilation aborted at op/taint.t line 15.
op/taint.t ........................ 
Dubious, test returned 255 (wstat 65280, 0xff00)
No subtests run 
Insecure dependency in require while running with -T switch at ./test.pl line 27.
BEGIN failed--compilation aborted at ./test.pl line 29.
Compilation failed in require at op/utftaint.t line 6.
BEGIN failed--compilation aborted at op/utftaint.t line 8.
op/utftaint.t ..................... 
Dubious, test returned 255 (wstat 65280, 0xff00)
No subtests run 
Insecure dependency in require while running with -T switch at ../../t/test.pl line 27.
BEGIN failed--compilation aborted at ../../t/test.pl line 29.
Compilation failed in require at t/DynaLoader.t line 9.
../ext/DynaLoader/t/DynaLoader.t .. 
Dubious, test returned 255 (wstat 65280, 0xff00)
No subtests run 
Insecure dependency in require while running with -T switch at ./test.pl line 27.
BEGIN failed--compilation aborted at ./test.pl line 29.
Compilation failed in require at perf/taint.t line 21.
BEGIN failed--compilation aborted at perf/taint.t line 23.
perf/taint.t ...................... 
Dubious, test returned 255 (wstat 65280, 0xff00)
No subtests run 

Test Summary Report
-------------------
op/taint.t                      (Wstat: 65280 Tests: 0 Failed: 0)
  Non-zero exit status: 255
  Parse errors: No plan found in TAP output
op/utftaint.t                   (Wstat: 65280 Tests: 0 Failed: 0)
  Non-zero exit status: 255
  Parse errors: No plan found in TAP output
../ext/DynaLoader/t/DynaLoader.t (Wstat: 65280 Tests: 0 Failed: 0)
  Non-zero exit status: 255
  Parse errors: No plan found in TAP output
perf/taint.t                    (Wstat: 65280 Tests: 0 Failed: 0)
  Non-zero exit status: 255
  Parse errors: No plan found in TAP output
Files=9, Tests=1099,  1 wallclock secs ( 0.06 usr  0.01 sys +  1.22 cusr  0.25 csys =  1.54 CPU)
Result: FAIL
```

The code in question in `t/test.pl`:
```
  22 BEGIN {
  23     my $lib = $^X;
  24     $lib =~ s{(\b)perl[^/]*$}{${1}lib};
  25 
  26     local @INC = ( $lib );
  27     require p5;
  28     p5->import;
  29 }
```

What needs to be fixed?

Thank you very much.
Jim Keenan

**Comment by atoomic** [comment]
I think we want to use a fullpath for `$lib` using `Cwd::abs_path`

**Comment by jkeenan** [comment]
I'm going to file one more bug related to `t/test.pl` later today.  So stay tuned.

**Comment by jkeenan** [comment]
> I'm going to file one more bug related to `t/test.pl` later today. So stay tuned.

https://github.com/atoomic/perl/issues/13 also pertains to `t/test.pl`.

**Comment by atoomic** [comment]
this is fixed by ba30ef11c6abfa87e2c2edac70bbf8c14c23e912

**Comment by jkeenan** [comment]
Looks good.
```
$ gitcurr
core-p7

$ git describe
v5.32.0-566-g2785301

$ ./perl -Ilib -V:config_args
config_args='-des -Dusedevel -Duseithreads';
...
$ cat taint-related-tests.output.txt 
cd t && (rm -f perl; /bin/ln -s ../perl perl)
./runtests choose
t/run/switcht ...................... ok
t/op/taint ......................... Useless use of join or string in void context at op/taint.t line 106.
Useless use of sprintf in void context at op/taint.t line 2077.
Useless use of subtraction (-) in void context at op/taint.t line 2084.
Useless use of concatenation (.) or string in void context at op/taint.t line 2271.
Useless use of index in void context at op/taint.t line 2453.
Useless use of join or string in void context at tmp2866B line 2.
Argument "x" isn't numeric in numeric bitwise or (|) at op/taint.t line 1512.
Argument "y" isn't numeric in numeric bitwise or (|) at op/taint.t line 1512.
Argument "x" isn't numeric in numeric bitwise or (|) at op/taint.t line 1515.
Argument "y" isn't numeric in numeric bitwise or (|) at op/taint.t line 1515.
Use of uninitialized value $_[0] in join or string at op/taint.t line 106, <$fh> chunk 1.
Use of uninitialized value $_[0] in join or string at op/taint.t line 106.
Use of uninitialized value $_[0] in join or string at op/taint.t line 106.
Use of uninitialized value $tainted in eval "string" at (eval 42) line 1.
Argument "foo" isn't numeric in numeric 1's complement (~) at op/taint.t line 2142.
Argument "foo" isn't numeric in numeric 1's complement (~) at op/taint.t line 2142.
Redundant argument in sprintf at op/taint.t line 2254.
Redundant argument in sprintf at op/taint.t line 2256.
Redundant argument in sprintf at op/taint.t line 2254.
Redundant argument in sprintf at op/taint.t line 2256.
Redundant argument in sprintf at op/taint.t line 2254.
Redundant argument in sprintf at op/taint.t line 2256.
Redundant argument in sprintf at op/taint.t line 2254.
Redundant argument in sprintf at op/taint.t line 2256.
Argument "<" isn't numeric in multiplication (*) at op/taint.t line 2305.
Argument "<" isn't numeric in multiplication (*) at op/taint.t line 2309.
Argument "<" isn't numeric in multiplication (*) at op/taint.t line 2313.
Argument "<" isn't numeric in multiplication (*) at op/taint.t line 2315.
Use of uninitialized value in index at op/taint.t line 2453.
Use of uninitialized value $x in concatenation (.) or string at op/taint.t line 2514.
Use of uninitialized value $_[0] in join or string at op/taint.t line 106.
Use of uninitialized value $_[0] in join or string at op/taint.t line 106.
Use of uninitialized value $_[0] in join or string at op/taint.t line 106.
ok
t/op/utftaint ...................... ok
t/../ext/DynaLoader/t/DynaLoader ... ok
t/perf/taint ....................... ok
All tests successful.
Elapsed: 0 sec
u=0.01  s=0.00  cu=0.32  cs=0.04  scripts=5  tests=1202
```
These are now mundane things to deal with.  Thanks for your rapid response.
